gravitational/teleport
# Add shim install script

This PR adds a couple of shim scripts as described in https://github.com/gravitational/cloud/issues/10230.

These scripts will be hosted in a static URL, and can be called like this `curl https://cdn.teleport.dev/install.sh | bash -s 16.2.0`, and will substitute the recommended script URL in https://goteleport.com/download/.

The scripts try to match the requested Teleport with the install script version, defaulting to a minimum when necessary.

Then, it downloads the install script (`assets/install-scripts/install.sh`) via curl or wget, verifies its checksum, and runs it.

The differences between the two scripts are:

- the URL (install-vx.y.z.sh/install-connect.vx.y.z)
- the Teleport Connect version doesn't take a second argument (`$TELEPORT_EDITION`)

Other than that, they're identical.

## Tests

I've tested it with the following setup. Let me know if you think the tests should be committed too.

<details>

<summary>setup</summary>

docker-compose.yaml, spins up a bunch of docker containers that run the install script and a script that verifies if the expected binaries are present in the container:

```yaml
services:
  #install.sh
  ubuntu-jammy-oss:
    image: ubuntu:22.04
    volumes:
      - type: bind
        source: ../install-shim.sh
        target: /install-shim.sh
      - type: bind
        source: ./test-shim.sh
        target: /test-shim.sh
    command: |
      bash -c 'apt-get update;
      apt-get install -y curl;
      bash /install-shim.sh 16.1.8;
      bash /test-shim.sh 16.1.8 oss'

  ubuntu-jammy-oss-wget:
    image: ubuntu:22.04
    volumes:
      - type: bind
        source: ../install-shim.sh
        target: /install-shim.sh
      - type: bind
        source: ./test-shim.sh
        target: /test-shim.sh
    command: |
      bash -c 'apt-get update;
      apt-get install -y wget;
      bash /install-shim.sh 16.1.8;
      bash /test-shim.sh 16.1.8 oss'

  ubuntu-jammy-enterprise-newer-version:
    image: ubuntu:22.04
    volumes:
      - type: bind
        source: ../install-shim.sh
        target: /install-shim.sh
      - type: bind
        source: ./test-shim.sh
        target: /test-shim.sh
    command: |
      bash -c 'apt-get update;
      apt-get install -y curl;
      bash /install-shim.sh 16.2.0 enterprise;
      bash /test-shim.sh 16.2.0 enterprise'

  ubuntu-jammy-enterprise-older-version:
    image: ubuntu:22.04
    volumes:
      - type: bind
        source: ../install-shim.sh
        target: /install-shim.sh
      - type: bind
        source: ./test-shim.sh
        target: /test-shim.sh
    command: |
      bash -c 'apt-get update;
      apt-get install -y curl;
      bash /install-shim.sh 14.3.30 enterprise;
      bash /test-shim.sh 14.3.30 enterprise'
  
  connect--ubuntu-jammy-enterprise-older-version:
    image: ubuntu:22.04
    platform: linux/amd64
    volumes:
      - type: bind
        source: ../install-connect-shim.sh
        target: /install-connect-shim.sh
      - type: bind
        source: ./test-shim.sh
        target: /test-shim.sh
    command: |
      bash -c 'apt-get update;
      apt-get install -y curl;
      bash /install-connect-shim.sh 14.3.30;
      bash /test-shim.sh 14.3.30 connect'
```

run-all-tests.sh, runs compose and checks output:

```bash
#!/bin/bash
set -e -o pipefail
docker compose up
if docker compose logs | grep "INSTALL_SCRIPT_TEST_FAILURE"; then
    echo "ONE OR MORE TESTS FAILED"
    exit 1
fi
echo "ALL TESTS COMPLETED SUCCESSFULLY"
```

test-shim.sm, asserts if the environment matches the expected after running the shim script:

```bash
#!/bin/bash
function test_teleport() {
    if ! type teleport; then
        echo "INSTALL_SCRIPT_TEST_FAILURE: teleport not found"
    fi
}

function test_tctl() {
    if ! type tctl; then
        echo "INSTALL_SCRIPT_TEST_FAILURE: tctl not found"
    fi
}

function test_tsh() {
    if ! type tsh; then
        echo "INSTALL_SCRIPT_TEST_FAILURE: tsh not found"
    fi
}

function test_tbot() {
    if ! type tctl; then
        echo "INSTALL_SCRIPT_TEST_FAILURE: tbot not found"
    fi
}

function test_upgrader() {
    if ! type teleport-upgrade; then
        echo "INSTALL_SCRIPT_TEST_FAILURE: upgrader not found"
    fi
}

function test_connect() {
    if ! type teleport-connect; then
        echo "INSTALL_SCRIPT_TEST_FAILURE: connect not found"
    fi
}

function test_version() {
    teleport version
    EXPECTED_VERSION=$1
    if ! teleport version | grep -q "$EXPECTED_VERSION"; then
        echo "INSTALL_SCRIPT_TEST_FAILURE: found '$(teleport version)' instead of $EXPECTED_VERSION"
        exit 1
    fi
}

echo "RUNNING TEST"

case $2 in
cloud | team)
    test_teleport
    test_tctl
    test_tsh
    test_tbot
    test_upgrader
    test_version "$1"
    ;;
oss)
    test_teleport
    test_tctl
    test_tsh
    test_tbot
    test_version "$1"
    ;;
enterprise)
    test_teleport
    test_tctl
    test_tsh
    test_tbot
    test_version "$1"
    ;;
connect)
    test_connect
    ;;
*)
    echo "INSTALL_SCRIPT_TEST_FAILURE: unsupported edition $EDITION"
    ;;
esac
```

</details>

Contributes to https://github.com/gravitational/cloud/issues/10230.
